src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -292,6 +311,146 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
         }
     }
 
+    private IKeyRing GetCurrentKeyRingCoreNew(DateTime utcNow, bool forceRefresh)
+    {
+        Debug.Assert(utcNow.Kind == DateTimeKind.Utc);
+
+        // The 99% and perf-critical case is that there is no task in-flight and the cached
+        // key ring is valid.  We do what we can to avoid unnecessary overhead (locking,
+        // context switching, etc) on this path.
+
+        CacheableKeyRing? existingCacheableKeyRing = null;
+
+        // Can we return the cached keyring to the caller?
+        if (!forceRefresh)
+        {
+            existingCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (CacheableKeyRing.IsValid(existingCacheableKeyRing, utcNow))
+            {
+                return existingCacheableKeyRing.KeyRing;
+            }
+        }
+
+        // If work kicked off by a previous caller has completed, we should use those results
+        // We check this outside the lock to reduce contention in the common case (no task).
+        // Logically, it would probably make more sense to check this before checking whether
+        // the cache is valid - there could be a newer value available - but keeping that path
+        // fast is more important.  The next forced refresh or cache expiration will cause the
+        // new value to be picked up.
+        var existingTask = Volatile.Read(ref _cacheableKeyRingTask);
+        if (existingTask is not null && existingTask.IsCompleted)
+        {
+            var taskKeyRing = GetKeyRingFromCompletedTask(existingTask, utcNow); // Throws if the task failed
```

Comment:
How is the exception handled in this case? I didn't see any handling in the invocation path for this method but might've missed something. Do we need to react differently to `Canceled` or `Faulted` status here?

Reply:
I was not 100% sure I understood the relationship between `IsCompleted` and faults/cancellation.  My goals was that, if the task wasn't still running, we would do whatever it would have done inline - return or throw.

Reply:
The answer to your question might just be that it hits whoever called Protect/Unprotect (as it would if the work were completed synchronously).